### ISSUE TITLE
Support getrandom() in the vdso

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#3362]
It looks like there's a plan to implement getrandom() in the vdso for Linux 6.11. When this happens (and I guess when glibc takes advantage of this), Shadow will need to handle this similarly to how it handles `__vdso_gettimeofday`, `__vdso_getcpu`, etc.

There's no reason to do anything yet before it's merged, but I figured it would be good to make an issue so that we remember to keep an eye on it.

https://lwn.net/Articles/980447/
https://www.phoronix.com/news/getrandom-vDSO-RNG-Linux-6.11
https://lore.kernel.org/lkml/20240703183115.1075219-1-Jason@zx2c4.com/

> This is back after a bit of a hiatus. In the last attempt to do this in
the beginning of 2023, I think we reached consensus on a few things --
the use case, the vDSO implementation and semantics, its integration
with libc, the test code and documentation, and so forth. It was
basically "ready to go". Almost. But there was a lingering issue that
bogged this down, which is that it demanded some new mm semantics that
weren't very popular.
>
> [...]
>
> Assuming this goes well, the plan would be to take this through my
random.git tree for 6.11. It's cooking in linux-next now.

**Comment by stevenengler** [comment]
And there's a glibc path that's worth tracking:

https://www.phoronix.com/news/Glibc-Review-getrandom-vDSO
https://inbox.sourceware.org/libc-alpha/20240730132440.1181819-1-Jason@zx2c4.com/

**Comment by stevenengler** [comment]
It appears to now be used by Go (on the master branch at least): https://www.phoronix.com/news/Goland-getrandom-vDSO

**Comment by stevenengler** [comment]
And has now been merged into glibc, with a likely release in February:

https://www.phoronix.com/news/glibc-getrandom-vDSO-Merged

> Thus with the Glibc 2.41 release due out around the start of February, getrandom vDSO will be utilized when running on recent Linux kernel versions

It's likely that we'll see this in distros starting in one of Fedora 42 (March), Ubuntu 25.04 (April), or Debian 13 (~July).

**Comment by sporksmith** [comment]
Looking through the [kernel patch](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7a3fad30fd8b4b5e370906b3c554f64026f56c2f) a bit ...

The interface is a little bit more complicated than the `getrandom` syscall:

```
+/**
+ * __cvdso_getrandom_data - Generic vDSO implementation of getrandom() syscall.
+ * @rng_info:		Describes state of kernel RNG, memory shared with kernel.
+ * @buffer:		Destination buffer to fill with random bytes.
+ * @len:		Size of @buffer in bytes.
+ * @flags:		Zero or more GRND_* flags.
+ * @opaque_state:	Pointer to an opaque state area.
+ * @opaque_len:		Length of opaque state area.
+ *
+ * This implements a "fast key erasure" RNG using ChaCha20, in the same way that the kernel's
+ * getrandom() syscall does. It periodically reseeds its key from the kernel's RNG, at the same
+ * schedule that the kernel's RNG is reseeded. If the kernel's RNG is not ready, then this always
+ * calls into the syscall.
+ *
+ * If @buffer, @len, and @flags are 0, and @opaque_len is ~0UL, then @opaque_state is populated
+ * with a struct vgetrandom_opaque_params and the function returns 0; if it does not return 0,
+ * this function should not be used.
+ *
+ * @opaque_state *must* be allocated by calling mmap(2) using the mmap_prot and mmap_flags fields
+ * from the struct vgetrandom_opaque_params, and states must not straddle pages. Unless external
+ * locking is used, one state must be allocated per thread, as it is not safe to call this function
+ * concurrently with the same @opaque_state. However, it is safe to call this using the same
+ * @opaque_state that is shared between main code and signal handling code, within the same thread.
+ *
+ * Returns:	The number of random bytes written to @buffer, or a negative value indicating an error.
+ */
+static __always_inline ssize_t
+__cvdso_getrandom_data(const struct vdso_rng_data *rng_info, void *buffer, size_t len,
+		       unsigned int flags, void *opaque_state, size_t opaque_len)
+{
```

But I *think* our implementation can ignore the `rng_info` buffer and just fall back to the syscall, as the last line of that function does:

```
+fallback_syscall:
+	return getrandom_syscall(orig_buffer, orig_len, flags);
+}
```

There's also another interface that looks like a wrapper for the above. Are *both* exposed to the userspace program?

```
+static __always_inline ssize_t
+__cvdso_getrandom(void *buffer, size_t len, unsigned int flags, void *opaque_state, size_t opaque_len)
+{
+	return __cvdso_getrandom_data(__arch_get_vdso_rng_data(), buffer, len, flags, opaque_state, opaque_len);
+}
```

There's a new mmap flag, which IIUC code using the above interfaces is required to use to allocate the state:

```
  First, it adds a new kind of mapping to mmap(2), MAP_DROPPABLE, which
  lets the kernel zero out pages anytime under memory pressure, which
  enables allocating memory that never gets swapped to disk but also
  doesn't count as being mlocked.
```

I think ultimately we don't want to pass that flag through natively; if nothing else it'd break determinism if the kernel can zero out the pages whenever it wants. I think we can safely drop the flag and do the native mmap as usual. (We also definitely wouldn't want it with the shadow memory mapper enabled, allowing the kernel to zero out memory that's mapped into shadow)

I think such `mmap`s will currently fail in shadow since it doesn't recognize the flag, which will cause ~all programs to fail under shadow once they're using this glibc+kernel, since libc will try to use it to allocate the vdso getrandom state. We could preemptively add "support" to just ignore this flag, but then the native `vdso` function calls will fail since the state wasn't mapped with the expected flag.

So, it might be worth trying to get this implemented preemptively to avoid complete breakage when this starts launching in distros. I suppose we could also yolo implement the actual vdso function(s) without testing. Our current unit tests that use libc's getrandom should test it once run on a system that supports this.

**Comment by sporksmith** [comment]
From some discussion with @stevenengler for posterity:

* It might be nice to put the implementation shim-side when doing this instead of just turning it into a `getrandom` syscall. It would be relatively straightforward to do so, and if applications expect `getrandom` to be in the vdso they might start using it directly more freely instead of just using it to seed a faster prng.
* Most rust prngs support no-std, so are suitable to run in the shim <https://rust-random.github.io/book/guide-rngs.html#basic-pseudo-random-number-generators-prngs>
* It might be nice to put our current host-wide PRNG in host-shared memory so that we can use it directly from the shim. This would require it to be `VirtualAddressSpaceIndependent`. Many are in practice (e.g. steven points out that  'Xoshiro256PlusPlus' is just a `[u64; 4]`), but this is an implementation detail that could change out from under us.
* It's probably fine to just use some initial pseudo-random data from "main" prng in shadow to seed a separate shim-side prng.
* https://github.com/shadow/shadow/pull/3542 in its current form has shadow send 8 pseudorandom bytes to the shim during initialization, which is used to directly overwrite random data in the auxvec. If we end up having a shim-side prng it'd probably make sense to fill the auxvec using that instead (and have shadow send an initial seed or state for that instead).

**Comment by stevenengler** [comment]
> It's probably fine to just use some initial pseudo-random data from "main" prng in shadow to seed a separate shim-side prng.

I agree that it's probably fine to have a separate shim-side prng. My only minor concern is that you might end up with some confusing behaviour caused by having two independent sources of randomness. For example getting 10 random bytes from `getrandom()`, which would be handled by the shim, would give different values from 10 random bytes from reading `/dev/urandom`, which would be handled by Shadow. And varying the number of bytes read from `getrandom()` wouldn't affect the values returned from future reads from `/dev/urandom`.

A contrived example might be someone who wants to vary the "seed" of only a single host rather than the entire simulation seed with `--seed`. You could do this by using a bash script like:

```bash
# "reseed" the host's rng by throwing away a bunch of random bytes
head -c 10000 /dev/urandom > /dev/null
exec ./my-application
```

But this would only affect parts of the simulation that are handled by shadow, not the shim. So `getrandom()` would still return the same results in `my-application` regardless of if you include the `head` command above, since `getrandom()` is on a different prng than `/dev/urandom`.

Anyways, I think this is all too contrived, and so I think having a separate prng in the shim is okay.

**Comment by stevenengler** [comment]
Do we have a determinism test for `getrandom()`? Looking at `src/test/determinism/test_determinism.c`, it looks like we compare the bytes read from `/dev/random` and `/dev/urandom`, but we don't test the libc wrappers like `getrandom()`.

So as part of this issue I think we'll also need to add determinism tests for `getrandom()`, and I think might as well test `random()` and `getentropy()` as well.

----------

In a debian sid container with glibc 4.41 and with a 6.13.7 kernel, I added the following test to `src/test/determinism/test_determinism.c`:

```c
static int _test_getrandom() {
    unsigned char buf[16] = {0};
    ssize_t num = getrandom(&buf, 16, 0);
    if (num != 16) {
        printf("getrandom() returned %ld != 16\n");
        return EXIT_FAILURE;
    }
    printf("getrandom():");
    for (int i = 0; i < 16; ++i) {
        printf(" %02x", buf[i]);
    }
    printf("\n");
    return EXIT_SUCCESS;
}
```

and the test still passed. Since shadow doesn't yet handle getrandom in the vdso, it seems like the test program is not handling `getrandom()` in the vdso. But I'm not sure why. Maybe this isn't actually part of glibc 4.41, even though [phoronix](https://www.phoronix.com/news/GNU-C-Library-glibc-2.41) claims that it is?

```text
$ /lib/x86_64-linux-gnu/libc.so.6 
GNU C Library (Debian GLIBC 2.41-6) stable release version 2.41.
```

**Comment by sporksmith** [comment]
> (Though why is the syscall running with 32 bytes when I requested 16?)

Maybe try doing multiple calls to `getrandom`? I wonder if this syscall is being used to bootstrap the VDSO PRNG

**Comment by sporksmith** [comment]
> There's a new mmap flag `[MAP_DROPPABLE]`, which IIUC code using the above interfaces is required to use to allocate the state:

I wondered if the glibc `mmap` call to set up this memory area might be failing, causing it to fall back to the syscall. Looking at the shadow code I *think* we'd currently end up allowing it and blindly passing it through, which we probably don't want. But at any rate, would also be interesting to look for `mmap` calls with `MAP_DROPPABLE` in the strace log

**Comment by stevenengler** [comment]
> > (Though why is the syscall running with 32 bytes when I requested 16?)
> 
> Maybe try doing multiple calls to `getrandom`? I wonder if this syscall is being used to bootstrap the VDSO PRNG

Yes, this appears to be exactly what it's doing.

```text
Thread 1 "test-determinis" hit Catchpoint 1 (call to syscall getrandom), 0x00007ffff7fc6348 in getrandom ()
(gdb) bt
#0  0x00007ffff7fc6348 in getrandom ()
#1  0x00007ffff7e041ef in getrandom_vdso (buffer=0x7fffffffeb60, length=16, flags=0, cancel=true) at ../sysdeps/unix/sysv/linux/getrandom.c:211
#2  0x000055555555554f in _test_getrandom () at /tmp/shadow/src/test/determinism/test_determinism.c:107
#3  0x0000555555555e02 in main (argc=1, argv=0x7fffffffecb8) at /tmp/shadow/src/test/determinism/test_determinism.c:279
```

If I add multiple `_test_getrandom` calls, and therefore multiple libc `getrandom()` calls, the strace looks like:

```text
write(1, "starting _test_getrandom()\n", 27) = 27
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_DROPPABLE|MAP_ANONYMOUS, -1, 0) = 0x7fe8f3b34000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fe8f3b33000
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
getrandom("\x68\xb4\x3c\x21\xdd\xb4\xcf\x03\xa3\xf3\x74\xa0\xa6\xad\x9e\xad\x6f\xd7\x50\x6a\xb2\xe5\x6f\x6f\xef\x22\x02\x61\xef\xfd\x3a\x0f", 32, 0) = 32
write(1, "getrandom(): 73 50 90 5f 98 b1 9"..., 61) = 61
write(1, "getrandom(): b9 07 cb ca b2 87 f"..., 61) = 61
write(1, "getrandom(): b0 61 c8 b9 27 ba 6"..., 61) = 61
write(1, "getrandom(): a7 dc 6d 32 8f 0e 6"..., 61) = 61
write(1, "_test_getrandom() passed\n", 25) = 25
```

> > There's a new mmap flag `[MAP_DROPPABLE]`, which IIUC code using the above interfaces is required to use to allocate the state:
>
> I wondered if the glibc `mmap` call to set up this memory area might be failing, causing it to fall back to the syscall. Looking at the shadow code I *think* we'd currently end up allowing it and blindly passing it through, which we probably don't want. But at any rate, would also be interesting to look for `mmap` calls with `MAP_DROPPABLE` in the strace log 

This also appears to be correct.

When we run this in shadow, shadow adds a libc preload, so our `getrandom()` call never calls the real glibc vdso call path.

https://github.com/shadow/shadow/blob/21433360658e62b255740eecd581c439d1c1f39c/src/lib/preload-libc/syscall_wrappers.c#L253-L255

If we set `--use-preload-libc false` for the `determinism1a` and `determinism1b` tests, the tests still pass. If we look at the shadow strace, we see:

```text
00:00:01.000000000 [tid 1000] rt_sigprocmask(0, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] mmap(<pointer>, 4096, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_DROPPABLE, 4294967295, 0) = <pointer> (EINVAL)
00:00:01.000000000 [tid 1000] rt_sigprocmask(2, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] getrandom(<pointer>, 16, 0) = 16
00:00:01.000000000 [tid 1000] rt_sigprocmask(0, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] mmap(<pointer>, 4096, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_DROPPABLE, 4294967295, 0) = <pointer> (EINVAL)
00:00:01.000000000 [tid 1000] rt_sigprocmask(2, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] getrandom(<pointer>, 16, 0) = 16
00:00:01.000000000 [tid 1000] rt_sigprocmask(0, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] mmap(<pointer>, 4096, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_DROPPABLE, 4294967295, 0) = <pointer> (EINVAL)
00:00:01.000000000 [tid 1000] rt_sigprocmask(2, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] getrandom(<pointer>, 16, 0) = 16
00:00:01.000000000 [tid 1000] rt_sigprocmask(0, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] mmap(<pointer>, 4096, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_DROPPABLE, 4294967295, 0) = <pointer> (EINVAL)
00:00:01.000000000 [tid 1000] rt_sigprocmask(2, <pointer>, <pointer>, 8) = 0
00:00:01.000000000 [tid 1000] getrandom(<pointer>, 16, 0) = 16
```

So it seems like libc is attempting to set up the vdso path, but the `mmap` returns `EINVAL` and then gives up and makes the syscall directly.
